tests/unit/package_managers/test_npm.py

```diff
@@ -0,0 +1,289 @@
+                },
+                {
+                    "package": {"name": "spam", "version": "3.0.0"},
+                    "dependencies": [{"name": "eggs", "version": "4.0.0"}],
+                    "package_lock_file": ProjectFile(
+                        abspath="/some/other/path", template="some other text"
+                    ),
+                },
+            ],
+            {
+                "components": [
+                    Component(name="foo", version="1.0.0"),
+                    Component(name="bar", version="2.0.0"),
+                    Component(name="spam", version="3.0.0"),
+                    Component(name="eggs", version="4.0.0"),
+                ],
+                "environment_variables": [],
+                "project_files": [
+                    ProjectFile(abspath="/some/path", template="some text"),
+                    ProjectFile(abspath="/some/other/path", template="some other text"),
+                ],
+            },
+            id="multiple_input_package",
+        ),
+    ],
+)
+@mock.patch("cachi2.core.package_managers.npm._resolve_npm")
+def test_fetch_npm_source(
+    mock_resolve_npm: mock.Mock,
+    npm_request: Request,
+    npm_input_packages: dict[str, str],
+    resolved_packages: Dict[str, Any],
+    request_output: Dict[str, List[Any]],
+):
+    """Test fetch_npm_source with different Request inputs."""
+    mock_resolve_npm.side_effect = resolved_packages
+    output = npm.fetch_npm_source(npm_request)
+    expected_output = RequestOutput.from_obj_list(
+        components=request_output["components"],
+        environment_variables=request_output["environment_variables"],
+        project_files=request_output["project_files"],
+    )
+
+    assert output == expected_output
+
+
+@mock.patch("pathlib.Path.exists")
+def test_resolve_npm_no_lock(mock_exists: mock.Mock):
+    """Test resolve_npm where npm-shrinkwrap.json or package-lock.json do not exist."""
+    mock_exists.return_value = False
+    expected_error = (
+        "The npm-shrinkwrap.json or package-lock.json file must be present for the npm "
+        "package manager"
+    )
+    with pytest.raises(PackageRejected, match=expected_error):
+        npm._resolve_npm(Path("some/package/dir"), Path("some/output/dir"))
+
+
+@pytest.mark.parametrize(
+    "package_lock_json, expected_output",
+    [
+        pytest.param(
+            {
+                "name": "foo",
+                "version": "1.0.0",
+                "lockfileVersion": 1,
+                "dependencies": {
+                    "bar": {
+                        "version": "2.0.0",
+                        "resolved": "https://some.registry.org/bar/-/bar-2.0.0.tgz",
+                    },
+                },
+            },
+            {
+                "package": {"name": "foo", "version": "1.0.0"},
+                "dependencies": [{"name": "bar", "version": "2.0.0"}],
+            },
+            id="npm_v1_lockfile",
```

Comment:
Should we add also local dependency to unit test for v1, where resolved is missing and version is filepath to package? https://github.com/cachito-testing/npm-cachi2-smoketest/blob/lockfile-v1/package-lock.json#L333 I am not sure if current code does not require presence of version in lock-file.

Reply:
Added :+1: